libs/utils/analysis/thermal_analysis.py

```diff
@@ -0,0 +1,243 @@
+# limitations under the License.
+#
+
+""" Thermal Analysis Module """
+
+import matplotlib.gridspec as gridspec
+import matplotlib.pyplot as plt
+import pandas as pd
+import pylab as pl
+import operator
+from trappy.utils import listify
+from devlib.utils.misc import memoized, list_to_mask, mask_to_list
+
+from analysis_module import AnalysisModule
+from plot_utils import plot_generic
+from trace import ResidencyTime, ResidencyData
+from bart.common.Utils import area_under_curve
+
+
+class ThermalAnalysis(AnalysisModule):
+    """
+    Support for plotting Thermal Analysis data
+
+    :param trace: input Trace object
+    :type trace: :mod:`libs.utils.Trace`
+    """
+
+###############################################################################
+# Analysis properties
+###############################################################################
+
+    @property
+    @memoized
+    def thermal_zones(self):
+        """
+        Get thermal zone ids that appear in the trace
+        """
+        df = self._dfg_trace_event('thermal_temperature')
+        return df["thermal_zone"].unique().tolist()
+
+    @property
+    @memoized
+    def cpufreq_cdevs(self):
+        df = self._dfg_trace_event('thermal_power_cpu_limit')
+        res = df['cpus'].unique().tolist()
+        return [mask_to_list(mask) for mask in res]
+
+###############################################################################
+# DataFrame Getter Methods
+###############################################################################
+
+    def _dfg_thermal_zone_temperature(self, ids=None):
+        """
+        Get the temperature readings of one or more thermal zone(s).
+
+        :param ids: The thermal zones to consider
+        :type ids: list(int)
+        """
+        df = self._dfg_trace_event('thermal_temperature')
+        df = df[['id', 'thermal_zone', 'temp']]
+
+        if ids is not None:
+            df = df[df.id.isin(ids)]
+
+        return df
+
+    def _dfg_cpufreq_cooling_state(self, cpus=None):
+        """
+        Get the cpufreq cooling device states
+
+        :param cpus: The cpus to consider
+        :type cpus: list(int)
+        """
+        df = self._dfg_trace_event('thermal_power_cpu_limit')
+        df = df[['cpus', 'freq', 'cdev_state']]
+
+        if cpus is not None:
+            # Find masks that match the requested CPUs
+            # This can include other CPUs
+            masks = self._matchingMasks(cpus)
+            df = df[df.cpus.isin(masks)]
+
+        return df
+
+    def _dfg_devfreq_cooling_state(self, devices=None):
+        """
+        Get the temperature readings of one or more thermal zone(s).
+
+        :param devices: The devfreq devices to consider
+        :type device: list(str)
+        """
+        df = self._dfg_trace_event('thermal_power_devfreq_limit')
+        df = df[['type', 'freq', 'cdev_state']]
+
+        if devices is not None:
+            df = df[df.type.isin(devices)]
+
+        return df
+
```

Comment:
I don't really see the utility of these `_dfg`s, they just return a pretty trival-to-acquire subset of the full DataFrame, right?

Reply:
To be honest I think only the cpufreq one is useful since we juggle cpulists & cpumasks around. For the other two, I figured a one-line accessor could always be nice to have, but if I'm wrong I can evince those.

Reply:
Oh yeah, actually the cpufreq one _is_ useful, and it seems fair to have the devfreq one for symmetry. I don't see the point of the temperature one, but it doesn't really hurt either 🤷‍♂️